src/plugins/interaction/ROSClockServer/ROSClockServer.cpp

```diff
@@ -73,16 +73,40 @@ bool ROSClockServer::init(std::map<std::string, std::string> &mission_params,
         ros::init(argc, NULL, "scrimmage", ros::init_options::NoSigintHandler);
     }
 
-    nh_ = std::make_shared<ros::NodeHandle>();
-    clock_pub_ = nh_->advertise<rosgraph_msgs::Clock>("/clock", 1);
+    // rosmaster check thread
+    check_rosmaster_thread_ = std::thread(&ROSClockServer::check_rosmaster, this);
+    running_ = true;
 
-    // Tell other nodes to use the simulated time from "/clock"
-    ros::param::set("/use_sim_time", true);
+    return true;
+}
 
-    // Publish the first time message
-    publish_clock_msg(time_->t());
+void ROSClockServer::close(double t) {
+    (void) t;
+    std::cout << "ROSClockServer: closing" << std::endl;
+    running_ = false;
+    check_rosmaster_thread_.join();
+}
 
-    return true;
+void ROSClockServer::check_rosmaster() {
+    while (running_) {
+        bool rosmaster_state = ros::master::check();
+        if (prev_rosmaster_state_ == false && rosmaster_state == true) {
+            // time to reinitialize
+            nh_ = std::make_shared<ros::NodeHandle>(); // this actually starts the node
+            // Create Publisher
+            pub_mutex_.lock();
+            clock_pub_.shutdown();
+            clock_pub_ = nh_->advertise<rosgraph_msgs::Clock>("/clock", 1);
+            ros::param::set("/use_sim_time", true);
+            pub_mutex_.unlock();
+            std::cout << "ROSClockServer: Ros initialized!" << std::endl;
```

Comment:
Are these `cout` statements necessary, or were they for debugging?  If only for debugging, please remove.